File: contains-duplicate/contains-duplicate.cpp
class Solution {
public:
    
    vector<int>& mergeSort(vector<int>& arr) {
        if(arr.size() > 1) {
            int m = arr.size() / 2; // m for midpoint 
            vector<int> L; 
            vector<int> R; 
            for (int i = 0; i < m; i++) {
                L.push_back(arr[i]);
            }
            for (int j = m; j < arr.size(); j++) {
                R.push_back(arr[j]); 
            }
            L = mergeSort(L);
            R = mergeSort(R);
            
            int i, j, k; 
            i = j = k = 0; 
            
            while ((i < L.size()) && (j < R.size())) {
                if (L[i] < R[j]){
                    arr[k] = L[i]; 
                    i++;
                    k++; 
                }
                else {
                    arr[k] = R[j];
                    j++; 
                    k++; 
                }
            }
            while (i < L.size()) {
                arr[k] = L[i]; 
                i++;
                k++; 
            }
            while (j < R.size()) {
                arr[k] = R[j]; 
                j++;
                k++; 
            }
            
        }
        return arr;   
    }
    
    bool containsDuplicate(vector<int>& nums) { 
        vector<int> sorted_nums = mergeSort(nums); 
        if (sorted_nums.size() == 1) return false; 
        int item_0 = sorted_nums[0];
        int item_1 = sorted_nums[1]; 
        for (int i = 1; i < sorted_nums.size(); i++) {
            item_0 = sorted_nums[i - 1]; 
            item_1 = sorted_nums[i]; 
            if (item_0 == item_1) return true; 
               
        }
        return false; 
    }
    
//     bool containsDuplicate(vector<int>& nums) {
//         int array_size = sizeof(nums) / sizeof(nums[0]); // can just say nums.size() 
//         for (int i = 0; i < array_size - 1; i++) {
//             for (int j = i + 1; j < array_size; j++) { 
//                 if (nums[i] == nums[j]) {
//                     return true; 
//                 }
//             }
//         }
//     return false; 
        
//     }
};

File: contains-duplicate/contains-duplicate.py
class Solution(object):
    
    def mergeSort(self, arr): # , l, r):
        if len(arr) > 1: 
            m = len(arr) // 2 
            L = arr[:m] 
            R = arr[m:] 
            self.mergeSort(L) #, 0, m)
            self.mergeSort(R) # , 0, len(R)) 
            
            i = j = k = 0 
            
            while i < len(L) and j < len(R):
                if L[i] < R[j]:
                    arr[k] = L[i] 
                    i += 1 
                    k += 1 
                else: 
                    arr[k] = R[j] 
                    j += 1 
                    k += 1 
            while i < len(L): 
                arr[k] = L[i] 
                i += 1 
                k += 1
            while j < len(R):
                arr[k] = R[j] 
                j += 1 
                k += 1    
        return arr 
        
    def containsDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        
        sorted_nums = self.mergeSort(nums) 
        print ("sorted array:")
        print (sorted_nums)
        
        if len(nums) == 1:
            return False 
        
        item_0 = sorted_nums[0] 
        item_1 = sorted_nums[1] 
        
        for ix, item_1 in enumerate(sorted_nums[1:]): 
            item_0 = sorted_nums[ix] 
            if item_1 == item_0:
                return True 
        return False 
        
        
        # for i in range(len(nums) - 1):
        #     for j in range(i + 1, len(nums)): 
        #         if nums[i] == nums[j]:
        #             return True 
        # return False 
        